carcara/src/checker/rules/pb_blasting.rs

```
@@ -399,8 +399,62 @@ fn get_bitvector_terms(bv: &Rc<Term>, pool: &mut dyn TermPool) -> Vec<Rc<Term>>
 }
 
 /// Implements the bitwise exclusive or operation.
-pub fn pbblast_bvxor(RuleArgs { .. }: RuleArgs) -> RuleResult {
-    Err(CheckerError::Unspecified)
+pub fn pbblast_bvxor(RuleArgs { pool, conclusion, .. }: RuleArgs) -> RuleResult {
+    let ((x, y), bit_constraints) =
+        match_term_err!((= (bvxor x y) (pbbterm ...)) = &conclusion[0])?;
+
+    let xs = get_bitvector_terms(x, pool);
+    let ys = get_bitvector_terms(y, pool);
+
+    // Zip three lists into tuples
+    for ((bc, xi), yi) in bit_constraints.iter().zip(xs.iter()).zip(ys.iter()) {
+        let (bindings, (c1, c2, c3, c4)) = match_term_err!((choice ... (and c1 c2 c3 c4)) = bc)?;
+
+        // Single binding
+        assert!(bindings.len() == 1);
+        // Check z -> Int
+        let (z_name, z_type) = &bindings[0];
+        assert!(z_name == "z");
+        assert!(*z_type.as_sort().unwrap() == Sort::Int);
```

One small problem: here, you are using `assert!`, instead of `rassert!`, which means this code will panic instead of returning a nice error message if the assertions are not valid.

For the first assertion, the parser already guarantees that there will be only one binding, so it's okay to use `assert!` (it's also okay to not check that at all, and just trust the parser). For the other two assertions, however, since they can be controlled by the user input, it would be better to use `rassert!`, and specify an error message.

This also affects the pull request for `pbblast_bvand` that was recently merged (my bad, I didn't catch that on the review), so please change that rule as well.

---

Does it look better now?